File: validation/python/corral_validate.py
"""
corral-validate: Minimal session validation for Corral/Better Auth.

Reads the shared Better Auth database directly. Zero dependencies beyond
stdlib (sqlite3) or psycopg2 for PostgreSQL.

## Auto-spawn auth server
    By default, CorralValidator will auto-spawn `node server/auth.js` as a
    managed subprocess so that login/signup endpoints are available. Session
    validation itself reads the DB directly and works without the auth server.
    Pass `auth_server=False` to skip (e.g. production/Docker where Node runs
    separately). Configure port via CORRAL_AUTH_PORT env var (default 3456)
    and server path via CORRAL_AUTH_SERVER env var.

## FastAPI
    from corral_validate import corral_auth_dependency
    auth = corral_auth_dependency("/data/auth.db")

    @app.get("/me")
    def me(user=Depends(auth)):
        return {"email": user.email}

## Flask
    from corral_validate import corral_auth_middleware
    corral_auth_middleware(app, "/data/auth.db")

    @app.route("/me")
    def me():
        return {"email": g.corral_user.email}

## Django (settings.py)
    MIDDLEWARE = ["corral_validate.CorralMiddleware", ...]
    CORRAL_DB_PATH = "/data/auth.db"
"""

from __future__ import annotations

import atexit
import logging
import os
import signal
import sqlite3
import subprocess
import threading
import time as _time
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
from urllib.request import urlopen
from urllib.error import URLError

__all__ = [
    "User", "CorralValidator", "corral_auth_dependency",
    "corral_auth_middleware", "CorralMiddleware",
]

PLAN_LEVELS = {"free": 0, "pro": 1, "team": 2, "enterprise": 3}
COOKIE_NAME = "better-auth.session_token"

SESSION_SQL = """
    SELECT s."userId", s."expiresAt"
    FROM "session" s WHERE s."token" = ?
"""
USER_SQL = """
    SELECT u."id", u."email", u."name", u."plan", u."role",
           u."emailVerified", u."createdAt"
    FROM "user" u WHERE u."id" = ?
"""

logger = logging.getLogger("corral_validate")


@dataclass
class User:
    id: str
    email: str
    name: Optional[str]
    plan: str
    role: str
    emailVerified: bool
    createdAt: str


class CorralValidator:
    """Validate Better Auth sessions against the shared database.

    Args:
        db_path_or_url: SQLite file path or PostgreSQL DSN.
        db_type: "sqlite" (default) or "postgres".
        auth_server: If True (default), auto-spawn the Node auth server
            as a managed subprocess. Set False in production/Docker where
            Node runs separately via supervisord etc.
    """

    def __init__(self, db_path_or_url: str, db_type: str = "sqlite", auth_server: bool = True):
        self.db_path = db_path_or_url
        self.db_type = db_type
        self._auth_proc: Optional[subprocess.Popen] = None
        self._auth_lock = threading.Lock()
        self._shutting_down = False

        if auth_server:
            self._start_auth_server()

    def _find_auth_server(self) -> Optional[str]:
        """Locate server/auth.js by env var or walking up from db_path."""
        env_path = os.environ.get("CORRAL_AUTH_SERVER")
        if env_path:
            return env_path if os.path.isfile(env_path) else None

        # Walk up from db_path looking for server/auth.js
        current = Path(self.db_path).resolve().parent
        for _ in range(10):
            candidate = current / "server" / "auth.js"
            if candidate.is_file():
                return str(candidate)
            parent = current.parent
            if parent == current:
                break
            current = parent
        return None

    def _start_auth_server(self) -> None:
        """Spawn the Node auth server as a managed subprocess."""
        auth_port = os.environ.get("CORRAL_AUTH_PORT", "3456")
        server_path = self._find_auth_server()

        if not server_path:
            logger.warning(
                "[corral-auth] server/auth.js not found — auth operations "
                "(login/signup) won't work, but session validation still works"
            )
            return

        # Check if node is available
        try:
            subprocess.run(["node", "--version"], capture_output=True, check=True)
        except (FileNotFoundError, subprocess.CalledProcessError):
            logger.warning(
                "[corral-auth] Node.js not installed — skipping auth server spawn"
            )
            return

        env = {**os.environ, "AUTH_PORT": auth_port}

        try:
            self._auth_proc = subprocess.Popen(
                ["node", server_path],
                env=env,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
            )
        except OSError as e:
            logger.warning("[corral-auth] Failed to spawn auth server: %s", e)
            return

        # Pipe stdout/stderr to logging in background threads
        def _pipe_log(stream, level):
            try:
                for line in stream:
                    text = line.decode("utf-8", errors="replace").rstrip()
                    if text:
                        logger.log(level, "[corral-auth] %s", text)
            except Exception:
                pass

        threading.Thread(
            target=_pipe_log, args=(self._auth_proc.stdout, logging.INFO),
            daemon=True, name="corral-auth-stdout",
        ).start()
        threading.Thread(
            target=_pipe_log, args=(self._auth_proc.stderr, logging.WARNING),
            daemon=True, name="corral-auth-stderr",
        ).start()

        # Health check
        url = f"http://localhost:{auth_port}/api/auth/ok"
        deadline = _time.monotonic() + 5.0
        healthy = False
        while _time.monotonic() < deadline:
            try:
                resp = urlopen(url, timeout=1)
                if resp.status == 200:
                    healthy = True
                    break
            except (URLError, OSError):
                pass
            _time.sleep(0.1)

        if healthy:
            logger.info("[corral-auth] Auth server ready on port %s (pid %d)", auth_port, self._auth_proc.pid)
        else:
            logger.warning("[corral-auth] Auth server health check failed after 5s — it may still be starting")

        # Register cleanup
        atexit.register(self._stop_auth_server)
        # Handle signals (only in main thread)
        try:
            original_sigterm = signal.getsignal(signal.SIGTERM)
            original_sigint = signal.getsignal(signal.SIGINT)

            def _handle_signal(signum, frame):
                self._stop_auth_server()
                # Call original handler
                orig = original_sigterm if signum == signal.SIGTERM else original_sigint
                if callable(orig) and orig not in (signal.SIG_DFL, signal.SIG_IGN):
                    orig(signum, frame)
                elif orig == signal.SIG_DFL:
                    signal.signal(signum, signal.SIG_DFL)
                    os.kill(os.getpid(), signum)

            signal.signal(signal.SIGTERM, _handle_signal)
            signal.signal(signal.SIGINT, _handle_signal)
        except ValueError:
            # Not in main thread — signals can't be set, atexit is enough
            pass

    def _stop_auth_server(self) -> None:
        """Gracefully shut down the auth server subprocess."""
        with self._auth_lock:
            if self._shutting_down or self._auth_proc is None:
                return
            self._shutting_down = True
            proc = self._auth_proc
            self._auth_proc = None

        if proc.poll() is not None:
            return

        logger.info("[corral-auth] Stopping auth server (pid %d)", proc.pid)
        try:
            proc.terminate()  # SIGTERM
            try:
                proc.wait(timeout=3)
            except subprocess.TimeoutExpired:
                proc.kill()  # SIGKILL
                proc.wait(timeout=2)
        except OSError:
            pass

    def _connect(self):
        if self.db_type == "postgres":
            try:
                import psycopg2
            except ImportError:
                raise RuntimeError("psycopg2 required for PostgreSQL support")
            return psycopg2.connect(self.db_path)
        return sqlite3.connect(self.db_path)

    def _query(self, sql: str, params: tuple):
        # PostgreSQL uses %s placeholders
        if self.db_type == "postgres":
            sql = sql.replace("?", "%s")
        conn = self._connect()
        try:
            cur = conn.cursor()
            cur.execute(sql, params)
            return cur.fetchone()
        finally:
            conn.close()

    def validate_session(self, token: str) -> Optional[User]:
        """Look up session by token, check expiry, return User or None."""
        row = self._query(SESSION_SQL, (token,))
        if not row:
            return None
        user_id, expires_at = row
        # Handle both string and datetime expiry
        if isinstance(expires_at, str):
            exp = datetime.fromisoformat(expires_at.replace("Z", "+00:00"))
        else:
            exp = expires_at if expires_at.tzinfo else expires_at.replace(tzinfo=timezone.utc)
        if exp < datetime.now(timezone.utc):
            return None
        return self.get_user_by_id(user_id)

    def get_user_by_id(self, user_id: str) -> Optional[User]:
        """Fetch a user by ID."""
        row = self._query(USER_SQL, (user_id,))
        if not row:
            return None
        return User(
            id=row[0], email=row[1], name=row[2],
            plan=row[3] or "free", role=row[4] or "user",
            emailVerified=bool(row[5]), createdAt=str(row[6]),
        )

    @staticmethod
    def require_plan(user: User, plan: str) -> bool:
        """Check if user's plan meets the minimum required plan."""
        return PLAN_LEVELS.get(user.plan, 0) >= PLAN_LEVELS.get(plan, 0)


def _extract_token(request) -> Optional[str]:
    """Extract token from cookie or Authorization header (generic)."""
    # Try cookie
    token = None
    cookies = getattr(request, "cookies", None) or {}
    if isinstance(cookies, dict):
        token = cookies.get(COOKIE_NAME)
    # Try Authorization header
    if not token:
        auth = None
        headers = getattr(request, "headers", {})
        if hasattr(headers, "get"):
            auth = headers.get("authorization") or headers.get("Authorization")
        if auth and auth.startswith("Bearer "):
            token = auth[7:]
    return token


# --- FastAPI ---

def corral_auth_dependency(db_path: str, db_type: str = "sqlite", auth_server: bool = True):
    """Return a FastAPI dependency that validates the session.

    Usage:
        auth = corral_auth_dependency("/data/auth.db")
        @app.get("/me")
        def me(user=Depends(auth)):
            return {"email": user.email}
    """
    validator = CorralValidator(db_path, db_type, auth_server=auth_server)

    async def dependency(request):
        from fastapi import HTTPException
        token = _extract_token(request)
        if not token:
            raise HTTPException(401, "No session token")
        user = validator.validate_session(token)
        if not user:
            raise HTTPException(401, "Invalid or expired session")
        return user

    return dependency


# --- Flask ---

def corral_auth_middleware(app, db_path: str, db_type: str = "sqlite", auth_server: bool = True):
    """Register a Flask before_request hook that sets g.corral_user.

    Usage:
        corral_auth_middleware(app, "/data/auth.db")
        # then in any route: g.corral_user
    """
    validator = CorralValidator(db_path, db_type, auth_server=auth_server)

    @app.before_request
    def _validate():
        from flask import request, g, abort
        token = _extract_token(request)
        if not token:
            g.corral_user = None
            return
        g.corral_user = validator.validate_session(token)


# --- Django ---

class CorralMiddleware:
    """Django middleware: sets request.corral_user.

    settings.py:
        MIDDLEWARE = ["corral_validate.CorralMiddleware", ...]
        CORRAL_DB_PATH = "/data/auth.db"
        CORRAL_DB_TYPE = "sqlite"  # optional
        CORRAL_AUTH_SERVER = False  # optional, default True
    """

    def __init__(self, get_response):
        self.get_response = get_response
        from django.conf import settings
        self.validator = CorralValidator(
            getattr(settings, "CORRAL_DB_PATH", "auth.db"),
            getattr(settings, "CORRAL_DB_TYPE", "sqlite"),
            auth_server=getattr(settings, "CORRAL_AUTH_SERVER", True),
        )

    def __call__(self, request):
        token = _extract_token(request)
        request.corral_user = self.validator.validate_session(token) if token else None
        return self.get_response(request)

File: validation/ruby/corral_validate.rb
# frozen_string_literal: true

# corral-validate: Minimal session validation for Corral/Better Auth.
#
# == Auto-spawn auth server
#   By default, pass `auth_server: true` to auto-spawn `node server/auth.js`
#   as a managed subprocess. Session validation reads the DB directly and
#   works without it. Configure port via CORRAL_AUTH_PORT (default 3456)
#   and server path via CORRAL_AUTH_SERVER env var.
#
# == Basic usage
#   v = Corral::Validator.new(db_path: "/data/auth.db", auth_server: true)
#   user = v.validate_session("tok_abc")
#   puts user.email if user
#
# == Rack middleware
#   use Corral::RackMiddleware, db_path: "/data/auth.db"
#   # env["corral.user"] will be set if authenticated
#
# == Rails concern
#   class ApplicationController < ActionController::Base
#     include Corral::AuthConcern
#     self.corral_db_path = "/data/auth.db"
#   end

require "sqlite3"
require "time"
require "net/http"
require "uri"

module Corral
  COOKIE_NAME = "better-auth.session_token"
  PLAN_LEVELS = { "free" => 0, "pro" => 1, "team" => 2, "enterprise" => 3 }.freeze

  User = Struct.new(:id, :email, :name, :plan, :role, :email_verified, :created_at, keyword_init: true)

  class Validator
    # @param db_path [String] Path to SQLite database
    # @param auth_server [Boolean] Auto-spawn Node auth server (default false)
    def initialize(db_path:, auth_server: false)
      @db_path = db_path
      @auth_pid = nil
      @auth_stopped = false
      @mutex = Mutex.new

      spawn_auth_server if auth_server
    end

    def validate_session(token)
      db = SQLite3::Database.new(@db_path)
      row = db.get_first_row(
        'SELECT "userId", "expiresAt" FROM "session" WHERE "token" = ?', [token]
      )
      return nil unless row

      user_id, expires_at = row
      exp = Time.parse(expires_at) rescue Time.iso8601(expires_at)
      return nil if exp < Time.now.utc

      get_user_by_id(db, user_id)
    ensure
      db&.close
    end

    def get_user_by_id(db = nil, user_id)
      own_db = db.nil?
      db ||= SQLite3::Database.new(@db_path)
      row = db.get_first_row(
        'SELECT "id","email","name","plan","role","emailVerified","createdAt" FROM "user" WHERE "id" = ?',
        [user_id]
      )
      return nil unless row

      User.new(
        id: row[0], email: row[1], name: row[2],
        plan: row[3] || "free", role: row[4] || "user",
        email_verified: !!row[5], created_at: row[6].to_s
      )
    ensure
      db&.close if own_db
    end

    def require_plan(user, plan)
      (PLAN_LEVELS[user.plan] || 0) >= (PLAN_LEVELS[plan] || 0)
    end

    # Stop the managed auth server subprocess.
    def stop_auth_server
      @mutex.synchronize do
        return if @auth_stopped || @auth_pid.nil?
        @auth_stopped = true
        pid = @auth_pid
        @auth_pid = nil

        begin
          # Check if still running
          Process.kill(0, pid)
          $stderr.puts "[corral-auth] Stopping auth server (pid #{pid})"
          Process.kill("TERM", pid)
          # Wait up to 3s
          deadline = Time.now + 3
          loop do
            _pid, _status = Process.waitpid2(pid, Process::WNOHANG)
            break if _pid
            break if Time.now > deadline
            sleep 0.05
          end
          # Force kill if still alive
          begin
            Process.kill(0, pid)
            Process.kill("KILL", pid)
            Process.waitpid(pid)
          rescue Errno::ESRCH, Errno::ECHILD
            # already gone
          end
        rescue Errno::ESRCH, Errno::ECHILD
          # already gone
        end
      end
    end

    private

    def spawn_auth_server
      port = ENV.fetch("CORRAL_AUTH_PORT", "3456")
      server_path = find_auth_server

      unless server_path
        $stderr.puts "[corral-auth] server/auth.js not found — auth operations won't work, session validation still works"
        return
      end

      # Check node is available
      unless system("node", "--version", out: File::NULL, err: File::NULL)
        $stderr.puts "[corral-auth] Node.js not installed — skipping auth server spawn"
        return
      end

      env = { "AUTH_PORT" => port }
      rd_out, wr_out = IO.pipe
      rd_err, wr_err = IO.pipe

      @auth_pid = Process.spawn(env, "node", server_path, out: wr_out, err: wr_err)
      wr_out.close
      wr_err.close

      # Pipe output to stderr with prefix in background threads
      pipe_to_log = ->(io) {
        Thread.new do
          io.each_line { |line| $stderr.puts "[corral-auth] #{line.rstrip}" unless line.strip.empty? }
        rescue IOError
          # closed
        end
      }
      pipe_to_log.call(rd_out)
      pipe_to_log.call(rd_err)

      # Health check
      url = URI("http://localhost:#{port}/api/auth/ok")
      deadline = Time.now + 5
      healthy = false
      while Time.now < deadline
        begin
          res = Net::HTTP.get_response(url)
          if res.code == "200"
            healthy = true
            break
          end
        rescue StandardError
          # not ready yet
        end
        sleep 0.1
      end

      if healthy
        $stderr.puts "[corral-auth] Auth server ready on port #{port} (pid #{@auth_pid})"
      else
        $stderr.puts "[corral-auth] Auth server health check failed after 5s — it may still be starting"
      end

      # Register cleanup
      at_exit { stop_auth_server }
      trap("TERM") { stop_auth_server; exit }
      trap("INT") { stop_auth_server; exit }
    rescue => e
      $stderr.puts "[corral-auth] Failed to spawn auth server: #{e.message}"
    end

    def find_auth_server
      if (p = ENV["CORRAL_AUTH_SERVER"])
        return File.file?(p) ? p : nil
      end
      dir = File.dirname(File.realpath(@db_path)) rescue File.dirname(@db_path)
      10.times do
        candidate = File.join(dir, "server", "auth.js")
        return candidate if File.file?(candidate)
        parent = File.dirname(dir)
        break if parent == dir
        dir = parent
      end
      nil
    end
  end

  # Rack middleware — sets env["corral.user"]
  class RackMiddleware
    def initialize(app, db_path:, auth_server: false)
      @app = app
      @validator = Validator.new(db_path: db_path, auth_server: auth_server)
    end

    def call(env)
      token = extract_token(env)
      env["corral.user"] = token ? @validator.validate_session(token) : nil
      @app.call(env)
    end

    private

    def extract_token(env)
      # Cookie
      cookies = env["HTTP_COOKIE"]&.split(";")&.map(&:strip) || []
      cookie = cookies.find { |c| c.start_with?("#{COOKIE_NAME}=") }
      return cookie.split("=", 2).last if cookie

      # Authorization header
      auth = env["HTTP_AUTHORIZATION"]
      return auth[7..] if auth&.start_with?("Bearer ")

      nil
    end
  end

  # Rails concern
  module AuthConcern
    def self.included(base)
      base.class_attribute :corral_db_path, default: "auth.db"
      base.class_attribute :corral_auth_server, default: false
      base.before_action :_corral_authenticate
    end

    private

    def _corral_authenticate
      validator = Validator.new(db_path: self.class.corral_db_path, auth_server: self.class.corral_auth_server)
      token = cookies[COOKIE_NAME] ||
              request.headers["Authorization"]&.then { |a| a.start_with?("Bearer ") ? a[7..] : nil }
      @corral_user = token ? validator.validate_session(token) : nil
    end

    def corral_user
      @corral_user
    end

    def require_corral_auth!
      head :unauthorized unless @corral_user
    end
  end
end
